neqo-transport/src/connection/tests/mod.rs

```diff
@@ -170,10 +193,29 @@ fn handshake(
         )
     };
 
+    let mut did_ping = enum_map! {_ => false};
     while !is_done(a) {
         _ = maybe_authenticate(a);
         let had_input = input.is_some();
+        // Insert a PING frame into the first application data packet an endpoint sends,
+        // in order to force the peer to ACK it. For the client, this is depending on
+        // tls_info() only returning something when the TLS handshake is complete.
+        // For the server, this is depending on the client's connection state, which is
+        // accessible during the tests.
+        //
+        // We're doing this to prevent packet loss from delaying ACKs, which would cause
+        // cwnd to shrink, which is not something the tests are written to account for.
+        let should_ping = !did_ping[a.role()]
+            && (a.role() == Role::Client && a.tls_info().is_some()
+                || (a.role() == Role::Server && b.state() == &State::Connected));
+        if should_ping {
+            a.test_frame_writer = Some(Box::new(Ping {}));
+        }
         let output = a.process(input.as_ref(), now).dgram();
+        if should_ping {
+            a.test_frame_writer = None;
+            did_ping[a.role()] = true;
```

Comment:
Let me see if I can work out your logic a bit better, because it doesn't seem like you need to ping on both peers for this to work.

I think that you can just have the server reach Confirmed and then force a PING into the packet it sends then.  The client will acknowledge that immediately, but have nothing further to send.

This arrangement has the client send a PING as well as soon as the connection is complete.  That happens when the state is Connected at the TLS layer, which corresponds with the Connected state for QUIC (I don't know why you need to check the TLS layer).  The client is already sending a TLS Finished as a Handshake packet at that point, which causes the server to send HANDSHAKE_DONE, so you don't need to do anything to get the server to respond immediately.

Reply:
What I saw was the client sending a bunch of new connection IDs to the server which remained unACK'ed (due to ACK delay being in effect). This change forces a PING into that packet, to trigger a server ACK.

Reply:
Ah, that's right.  I forget that I added that feature.  Won't it be enough for the client to do that then?  You can want for Confirmed and send a ping with that.

Reply:
I'll try. (I think i tried to do it client-only and there were some more test failures.)

Reply:
I don't think that works, because the `HandshakeDone` doesn't elicit an immediate ACK. So when the `handshake()` methid is done, the client still has the delayed ACK timer pending and is not idle. 

There is a case to be made to send an immediate ACK on `HandshakeDone`, which would fix this.

Reply:
(Well, it would fix it for a subset of tests, but many are still failing with a client-only PING and an immediate ACK on HandshakeDone.)

Reply:
OK to merge this as-is?

Reply:
Ah yes.  It's a bit ugly having the check for "Connected" on the client always, but I think you have it right.  We can't rely on the server Connected signal and the server Confirmed happens after the ping.  Thanks for checking.